File: js_validation/pardox/src/_io/databases.js
'use strict';

const { getLib } = require('../ffi');
const { DataFrame } = require('../DataFrame');

/**
 * Read a MySQL table/query into a DataFrame.
 * @param {string} connectionString  MySQL URL (e.g. "mysql://user:pass@host:3306/db").
 * @param {string} query             SQL query.
 * @returns {DataFrame}
 */
function read_mysql(connectionString, query) {
    return DataFrame.read_mysql(connectionString, query);
}

/**
 * Execute DDL/DML against MySQL.
 * @param {string} connectionString
 * @param {string} query
 * @returns {number} Rows affected.
 */
function execute_mysql(connectionString, query) {
    const lib  = getLib();
    const rows = lib.pardox_execute_mysql(connectionString, query);
    if (rows < 0) throw new Error(`execute_mysql failed with error code: ${rows}`);
    return rows;
}

/**
 * Read from SQL Server into a DataFrame.
 * @param {string} connectionString  ADO.NET connection string.
 * @param {string} query             SQL query.
 * @returns {DataFrame}
 */
function read_sqlserver(connectionString, query) {
    return DataFrame.read_sqlserver(connectionString, query);
}

/**
 * Execute DDL/DML against SQL Server.
 * @param {string} connectionString
 * @param {string} query
 * @returns {number} Rows affected.
 */
function execute_sqlserver(connectionString, query) {
    const lib  = getLib();
    const rows = lib.pardox_execute_sqlserver(connectionString, query);
    if (rows < 0) throw new Error(`execute_sqlserver failed with error code: ${rows}`);
    return rows;
}

/**
 * Read a MongoDB collection into a DataFrame.
 * @param {string} connectionString   MongoDB URI.
 * @param {string} dbDotCollection    e.g. "mydb.orders".
 * @returns {DataFrame}
 */
function read_mongodb(connectionString, dbDotCollection) {
    return DataFrame.read_mongodb(connectionString, dbDotCollection);
}

/**
 * Execute a MongoDB command.
 * @param {string} connectionString  MongoDB URI.
 * @param {string} database          Target database name.
 * @param {string} commandJson       JSON string of the command document.
 * @returns {number}
 */
function execute_mongodb(connectionString, database, commandJson) {
    const lib    = getLib();
    const result = lib.pardox_execute_mongodb(connectionString, database, commandJson);
    if (result < 0) throw new Error(`execute_mongodb failed with error code: ${result}`);
    return result;
}

/**
 * Check SQL Server driver configuration.
 * @returns {boolean}
 */
function sqlserverConfigOk() {
    const lib = getLib();
    return Boolean(lib.pardox_sqlserver_config_ok());
}

module.exports = {
    read_mysql,
    execute_mysql,
    read_sqlserver,
    execute_sqlserver,
    read_mongodb,
    execute_mongodb,
    sqlserverConfigOk,
};

File: py_validation/pardox/_io/databases.py
from ..wrapper import lib


def read_mysql(connection_string: str, query: str):
    """
    Reads data from a MySQL database using PardoX's native Rust driver.

    Args:
        connection_string (str): MySQL URL (e.g. "mysql://user:pass@host:3306/db")
        query (str): SQL query string.

    Returns:
        pardox.DataFrame
    """
    from ..frame import DataFrame
    if not hasattr(lib, 'pardox_read_mysql'):
        raise NotImplementedError("API 'pardox_read_mysql' not found in Core. Re-compile Rust.")
    conn_bytes  = connection_string.encode('utf-8')
    query_bytes = query.encode('utf-8')
    manager_ptr = lib.pardox_read_mysql(conn_bytes, query_bytes)
    if not manager_ptr:
        raise RuntimeError("MySQL query failed. Check console/stderr for Rust driver errors.")
    return DataFrame(manager_ptr)


def execute_mysql(connection_string: str, query: str) -> int:
    """
    Execute arbitrary DDL or DML directly against MySQL via the Rust native driver.

    Args:
        connection_string (str): MySQL URL.
        query (str): SQL statement to execute.

    Returns:
        int: Number of rows affected.
    """
    if not hasattr(lib, 'pardox_execute_mysql'):
        raise NotImplementedError("API 'pardox_execute_mysql' not found in Core. Re-compile Rust.")
    conn_bytes  = connection_string.encode('utf-8')
    query_bytes = query.encode('utf-8')
    rows = lib.pardox_execute_mysql(conn_bytes, query_bytes)
    if rows < 0:
        raise RuntimeError(f"execute_mysql failed with code: {rows}")
    return int(rows)


def read_sqlserver(connection_string: str, query: str):
    """
    Reads data from SQL Server using PardoX's native Rust driver (tiberius).

    Args:
        connection_string (str): ADO.NET connection string.
            e.g. "Server=localhost,1433;Database=mydb;User Id=sa;Password=pwd;TrustServerCertificate=true"
        query (str): SQL query string.

    Returns:
        pardox.DataFrame
    """
    from ..frame import DataFrame
    if not hasattr(lib, 'pardox_read_sqlserver'):
        raise NotImplementedError("API 'pardox_read_sqlserver' not found in Core. Re-compile Rust.")
    conn_bytes  = connection_string.encode('utf-8')
    query_bytes = query.encode('utf-8')
    manager_ptr = lib.pardox_read_sqlserver(conn_bytes, query_bytes)
    if not manager_ptr:
        raise RuntimeError("SQL Server query failed. Check console/stderr for Rust driver errors.")
    return DataFrame(manager_ptr)


def execute_sqlserver(connection_string: str, query: str) -> int:
    """
    Execute arbitrary DDL or DML directly against SQL Server via the Rust native driver.

    Args:
        connection_string (str): ADO.NET connection string.
        query (str): SQL statement to execute.

    Returns:
        int: Number of rows affected.
    """
    if not hasattr(lib, 'pardox_execute_sqlserver'):
        raise NotImplementedError("API 'pardox_execute_sqlserver' not found in Core. Re-compile Rust.")
    conn_bytes  = connection_string.encode('utf-8')
    query_bytes = query.encode('utf-8')
    rows = lib.pardox_execute_sqlserver(conn_bytes, query_bytes)
    if rows < 0:
        raise RuntimeError(f"execute_sqlserver failed with code: {rows}")
    return int(rows)


def read_mongodb(connection_string: str, db_dot_collection: str):
    """
    Reads a MongoDB collection into a PardoX DataFrame using the native Rust driver.

    Args:
        connection_string (str): MongoDB URI (e.g. "mongodb://user:pass@host:27017")
        db_dot_collection (str): Target as "database.collection" (e.g. "mydb.orders")

    Returns:
        pardox.DataFrame
    """
    from ..frame import DataFrame
    if not hasattr(lib, 'pardox_read_mongodb'):
        raise NotImplementedError("API 'pardox_read_mongodb' not found in Core. Re-compile Rust.")
    conn_bytes   = connection_string.encode('utf-8')
    target_bytes = db_dot_collection.encode('utf-8')
    manager_ptr = lib.pardox_read_mongodb(conn_bytes, target_bytes)
    if not manager_ptr:
        raise RuntimeError("MongoDB read failed. Check console/stderr for Rust driver errors.")
    return DataFrame(manager_ptr)


def execute_mongodb(connection_string: str, database: str, command_json: str) -> int:
    """
    Execute a MongoDB command (e.g. drop, create, custom commands) via the Rust native driver.

    Args:
        connection_string (str): MongoDB URI.
        database (str): Target database name.
        command_json (str): JSON string of the MongoDB command document.
            e.g. '{"drop": "my_collection"}'

    Returns:
        int: Value of the 'n' or 'ok' field from the command result.
    """
    if not hasattr(lib, 'pardox_execute_mongodb'):
        raise NotImplementedError("API 'pardox_execute_mongodb' not found in Core. Re-compile Rust.")
    conn_bytes = connection_string.encode('utf-8')
    db_bytes   = database.encode('utf-8')
    cmd_bytes  = command_json.encode('utf-8')
    result = lib.pardox_execute_mongodb(conn_bytes, db_bytes, cmd_bytes)
    if result < 0:
        raise RuntimeError(f"execute_mongodb failed with code: {result}")
    return int(result)


def sqlserver_config_ok() -> bool:
    """
    Check whether the SQL Server native driver is correctly configured.

    Returns:
        bool: True if configured, False otherwise.
    """
    if not hasattr(lib, 'pardox_sqlserver_config_ok'):
        raise NotImplementedError("pardox_sqlserver_config_ok not found in Core.")
    return bool(lib.pardox_sqlserver_config_ok())
